### ISSUE TITLE
How to disable empty GT filtering in a custom dataset of type COCODataset

### ISSUE THREAD

**Issue by AndreaPi** [open-mmlab/mmdetection#6215]
Hi,

I would like to train on all the images in my custom type `COCODataset` dataset. The dataset also includes images without labels, i.e., empty. Similarly, I would like to test on all the test set images, including the empty ones. By default, MMDetection filters out empty GT images. According to this PR, it should be possible to change this default

https://github.com/open-mmlab/mmdetection/pull/3695

but I couldn't find in the documentation how to modify the config file in order to disable empty GT filtering. Can you help me?

**Comment by Kaze-wind** [comment]
maybe you can use it like this:
```python
dataset=dict(
        type='CocoDataset',
        ann_file=<path_to_your_json>,
        img_prefix=<path_to_your_data_dir>,
        pipeline=[
            dict(type='LoadImageFromFile', to_float32=True),
            dict(type='LoadAnnotations', with_bbox=True)
        ],
        filter_empty_gt=True,
    )
```

**Comment by AndreaPi** [comment]
hmmm, seems to ignore it. Do you know how I can get the number of images being processed in each training epoch?

**Comment by Kaze-wind** [comment]
> hmmm, seems to ignore it. Do you know how I can get the number of images being processed in each training epoch?

The `samples_per_gpu` in the `data = dict(...)` config is the batch_size. Number in the log here `INFO - Epoch [1][250/438]` means `[present_iter/max_iter]`. So the number of train images should be `samples_per_gpu` x `max iter`.

The batch_size for valid / test seems to be 1, so the `max_iter` is the number.

**Comment by AndreaPi** [comment]
First of all, thanks for the support. Here's the relevant part of my config file:

```
# dataset settings
data = dict(
    samples_per_gpu=4,
    workers_per_gpu=2,
    type='CocoDataset',
    filter_empty_gt=False,
    train=dict(
        ann_file=<..>,
        img_prefix=<..>,
        pipeline=train_pipeline,
        classes=classes),
    val=dict(
        ann_file=<..>,
        img_prefix=<..>,
        pipeline=test_pipeline,
        classes=classes),
    test=dict(
        ann_file=<..>,
        img_prefix=<..>,
        pipeline=test_pipeline,
        classes=classes),
)
```

`samples_per_gpu` is equal to 4, and I have 1 GPU (as confirmed by `torch.cuda.device_count()`) ⇒ `batch_size = samples_per_gpu*num_gpus = 4`. Now, from my log:

`2021-10-02 15:31:22,460 - mmdet - INFO - Epoch [1][50/283]`

(note that there's no `Epoch(val)` qualifier, so this is a legit train epoch, not a val epoch). Thus `n_train_images = samples_per_gpu * max_iter * num_gpus = 4* 283*1=1132`

_However_, if I have a look at the json file of my train dataset, I see that 1208 images are included. So, it looks like 1208-1132=76 images are missing! And that would square with the number of empty GT images, so....

**Comment by AndreaPi** [comment]
BTW, here's another proof that MMDetection is still filtering empty images. I wrote a script to see how many images are in the train, val, test set which MMDetection builds:

```
from pathlib import Path
import mmcv
from mmcv import Config
from mmdet.datasets.builder import build_dataset

config_path = Path(...)

cfg = Config.fromfile(config_path)

train_dataset = build_dataset(cfg.data.train)
print(train_dataset)
val_dataset = build_dataset(cfg.data.val)
print(val_dataset)
test_dataset = build_dataset(cfg.data.test)
print(test_dataset)
```

The output is:

```
loading annotations into memory...
Done (t=0.06s)
creating index...
index created!

CocoDataset Train dataset with number of images 1129, and instance counts: 
+----------+-------+----------+-------+----------+-------+----------+-------+----------+-------+
| category | count | category | count | category | count | category | count | category | count |
+----------+-------+----------+-------+----------+-------+----------+-------+----------+-------+
| 0 [a]    | 798   | 1 [b]    | 569   | 2 [c]    | 286   | 3 [d]    | 406   | 4 [e]    | 211   |
+----------+-------+----------+-------+----------+-------+----------+-------+----------+-------+
loading annotations into memory...
Done (t=0.01s)
creating index...
index created!

CocoDataset Train dataset with number of images 142, and instance counts: 
+----------+-------+----------+-------+----------+-------+----------+-------+----------+-------+
| category | count | category | count | category | count | category | count | category | count |
+----------+-------+----------+-------+----------+-------+----------+-------+----------+-------+
| 0 [a]    | 110   | 1 [b]    | 73    | 2 [c]    | 41    | 3 [d]    | 53    | 4 [e]    | 16    |
+----------+-------+----------+-------+----------+-------+----------+-------+----------+-------+
loading annotations into memory...
Done (t=0.01s)
creating index...
index created!

CocoDataset Train dataset with number of images 142, and instance counts: 
+----------+-------+----------+-------+----------+-------+----------+-------+----------+-------+
| category | count | category | count | category | count | category | count | category | count |
+----------+-------+----------+-------+----------+-------+----------+-------+----------+-------+
| 0 [a]    | 89    | 1 [b]    | 83    | 2 [c]    | 29    | 3 [d]    | 60    | 4 [e]    | 31    |
+----------+-------+----------+-------+----------+-------+----------+-------+----------+-------+
```

So clearly the train dataset being loaded by MMDetection has only 1129 images, while it should have 1208!

**Comment by Kaze-wind** [comment]
Try this config ? At least it works for me. 
The `filter_empty_gt` outside the `train/val` seems invalid. So as `type`...
```python
data = dict(
    samples_per_gpu=4,
    workers_per_gpu=2,
    ## not declare here
    #type='CocoDataset',
    #filter_empty_gt=False,
    train=dict(
        #declare here
        type='CocoDataset',
        filter_empty_gt=False,
        # end
        ann_file=<..>,
        img_prefix=<..>,
        pipeline=train_pipeline,
        classes=classes),
    val=dict(
        type='CocoDataset',
        filter_empty_gt=False,
        ann_file=<..>,
        img_prefix=<..>,
        pipeline=test_pipeline,
        classes=classes),
    test=dict(
        type='CocoDataset',
        filter_empty_gt=False,
        ann_file=<..>,
        img_prefix=<..>,
        pipeline=test_pipeline,
        classes=classes),
)
```

**Comment by AndreaPi** [comment]
Works perfectly! Thank you @Kaze-wind . Closing issue now!
